CUDADataFormats/SiPixelCluster/interface/SiPixelClustersCUDA.h

```diff
@@ -5,16 +5,29 @@
 #include "HeterogeneousCore/CUDAUtilities/interface/host_unique_ptr.h"
 #include "HeterogeneousCore/CUDAUtilities/interface/cudaCompat.h"
 
+#include "DataFormats/SoATemplate/interface/SoALayout.h"
+#include "CUDADataFormats/Common/interface/PortableDeviceCollection.h"
+
 #include <cuda_runtime.h>
 
-class SiPixelClustersCUDA {
+GENERATE_SOA_LAYOUT(SiPixelClustersCUDALayout,
+                    SOA_COLUMN(uint32_t, moduleStart),
+                    SOA_COLUMN(uint32_t, clusInModule),
+                    SOA_COLUMN(uint32_t, moduleId),
+                    SOA_COLUMN(uint32_t, clusModuleStart))
+
+using SiPixelClustersCUDASoA = SiPixelClustersCUDALayout<>;
+using SiPixelClustersCUDASOAView = SiPixelClustersCUDALayout<>::View;
+using SiPixelClustersCUDASOAConstView = SiPixelClustersCUDALayout<>::ConstView;
+
+class SiPixelClustersCUDA : public cms::cuda::PortableDeviceCollection<SiPixelClustersCUDALayout<>> {
```

Comment:
This idea has been passed down to from you to us via @AdrianoDee. Should it be addressed in this PR? As you can see, it will not be a trivial change.

Reply:
If this code would be intended for longer term, I'd insist on addressing it in this PR. But given the temporal nature of these classes (being a stepping stone towards Alpaka-based implementation), and that all these data formats are transient, maybe the inheritance approach could be tolerated. I'd anyway suggest to add a comment for each of these classes that they should be using composition instead, to be done for the Alpaka versions in the near future, in case anyone would use these as an example in the mean time.

Reply:
I understand the inheritance makes some things easy. Nevertheless, a general guideline for class design is to prefer composition over inheritance when the latter is not necessary. There are lot's of material about that around the internet, e.g. here is the take of the C++ Core Guidelines https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-domain.

I personally see `Portable*Collection` conceptually similar to `std::vector`, and inheriting from `std::vector` is generally a bad idea.

In principle the `SiPixelClustersCUDA` could define its own `operator->()` function that would delegate the call to the View.

Although one could also ask if the `Portable*Collection` should give access to the columns, or only to the buffer and View.

Reply:
Most of the discussion about inheritance vs composition revolves around the use of virtual functions, pure virtual functions, and virtual destructors.
But those caveats do not apply here: we do not have virtual functions, and the derived types are the only ones that are going to be used.

> Although one could also ask if the `Portable*Collection` should give access to the columns, or only to the buffer and View.

Technically, the `Portable*Collection` _does_ give access only to the buffer and view.
Its `*` and `->` operators return a `View&` or `View*`, and access to the columns is done by that.

Reply:
> In principle the `SiPixelClustersCUDA` could define its own `operator->()` function that would delegate the call to the View.

Yes, I agree that this should be simple enough.

Reply:
> Most of the discussion about inheritance vs composition revolves around the use of virtual functions, pure virtual functions, and virtual destructors.
> But those caveats do not apply here: we do not have virtual functions, and the derived types are the only ones that are going to be used.

They kind of do, but in a different way. A class hierarchy whose members behave differently by overloading virtual functions is clearly a use case that necessitates inheritance. For the case where there are no virtual functions (or other reasons that would necessitate inheritance), the C++ core guidelines recommend
> Do _not_ use inheritance when simply having a data member will do.

For example, public inheritance allows code like below, even if it behaves incorrectly (and thus it adds cost for code review to spot such uses)
```cpp
std::unique_ptr<cms::cuda::PortableDeviceCollection<SiPixelClustersCUDALayout<>> foo = std::make_unique<SiPixelClustersCUDA>();
```

Another example of potentially unexpected behavior could be code like
```cpp
// overload for generic PortableDeviceCollection
template <typename T>
auto copy(cms::cuda::PortableDevicecollection<T> const& collDev, cudaStream_t stream) {
  cms::cuda::PortableHostCollection<T> collHost{collDev->metadata().size(), stream};
  cudaCheck(cudaMemcpyAsync(collHost.buffer().get(), collDev.buffer().get(), collDev.bufferSize(), cudaMemcpyDefault, stream);
  return collHost;
}
// oops, forgot to overload SiPixelClustersCUDA

SiPixelClustersCUDA objDevice;
auto objHost = copy(objDevice, stream);
```

Shortcuts in a limited set of code that is fully in one's control is kind-of cheap to tolerate (if the structure turns out to be a bad choice, changing the code should be fairly straightforward). But this case looks like we would be setting a pattern that could eventually spread to tens to hundreds of classes. I think we should be careful in such a situation, and keep in mind that users tend to find all kinds of loopholes in library design.